src/lib.rs

```diff
@@ -129,29 +146,134 @@ macro_rules! case {
     ($($variant:ident)::+) => {
         $crate::filter(|x| matches!(x, $($variant)::+))
     };
-    ($($variant:ident)::+ ($param:ident)) => {
+    ($($entry:tt)*) => {
         $crate::filter_map(|x| match x {
-            $($variant)::+($param) => Some($param),
+            $($entry)*
+            => {
+                #[allow(unused_imports)]
+                use ::tuples::combin::CombinConcat;
+                Some($crate::_extract_values!($($entry)*))
+            },
             _ => None,
         })
     };
-    ($($variant:ident)::+ ($($param:ident),+ $(,)?)) => {
-        $crate::filter_map(|x| match x {
-            $($variant)::+($($param),+) => Some(($($param),+ ,)),
-            _ => None,
-        })
+}
+
+/// A helper macro for [`case!`] to extract values recursively from structs and enums and returns a
+/// tuple with all of the matched values.
+///
+/// [`case!`]: crate::case
+//
+// This looks confusing, so here is how this works:
+//
+// It's just a recursive TT muncher that takes all of the "endpoint" values in rusts pattern
+// matching syntax.
+//
+// The first rule is just for `..` support.
+// differ because ".." can be at the front, where a lack of a value is an error, but a
+// comma indicates that the value before is filled in, and we can return nothing.
+//
+// 3-5 rules are escapes from the recursion.
+// All branches end in either a variable or an enum variant with no values (for filtering reasons).
+//
+// 6-7 rules feed the inners of structs and enums back to the macro, so that rules 8-13 can munch
+// on them.
+//
+// Speaking of which, 8-13 rules just take comma separated arguments and make them into a tuple.
+//
+// 14-15 rules just expand `key: value` to `value`
+#[macro_export]
+macro_rules! _extract_values {
```

Comment:
This macro is private, if I understand it correctly. But it's shown in the public documentation, which you can confirm by running `cargo doc --open`. Let's make it not appear in the documentation via [`#[doc(hidden)]`](https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#hidden).

Reply:
Done.